reccmp/tools/project.py

```diff
@@ -0,0 +1,457 @@
+                add_executable_or_library = "add_executable"
+                maybe_shared = ""
+            case TargetType.SHARED_LIBRARY:
+                add_executable_or_library = "add_library"
+                maybe_shared = "SHARED"
+        result += "\n"
+        result += textwrap.dedent(
+            f"""\
+            {add_executable_or_library}({target_name} {maybe_shared}
+                main_{target_name}.cpp
+                main_{target_name}.hpp
+            )
+            reccmp_add_target({target_name} ID {target_name})
+            set_property(TARGET {target_name} PROPERTY OUTPUT_NAME "{target_path.stem}")
+            set_property(TARGET {target_name} PROPERTY PREFIX "{target_prefix}")
+            set_property(TARGET {target_name} PROPERTY SUFFIX "{target_suffix}")
+        """
+        )
+
+    result += "\n"
+    result += textwrap.dedent(
+        """\
+        reccmp_configure()
+    """
+    )
+    return result
+
+
+def get_default_main_hpp(target_id: str) -> str:
+    return textwrap.dedent(
+        f"""\
+        #ifndef {target_id.upper()}_HPP
+        #define {target_id.upper()}_HPP
+
+        // VTABLE: {target_id} 0x10001000
+        // SIZE 0x8
+        class SomeClass {{
+            virtual ~SomeClass(); // vtable+0x00
+            int m_member;
+        }};
+
+        #endif /* {target_id.upper()}_HPP */
+        """
+    )
+
+
+def get_default_main_cpp(target_id: str, original_path: Path, hpp_path: Path) -> str:
+    target_type = executable_or_library(original_path)
+    match target_type:
+        case TargetType.EXECUTABLE:
+            entry_function = textwrap.dedent(
+                """\
+                #ifdef _WIN32
+                #include <windows.h>
+
+                // FUNCTION: {original_id} 0x10000020
+                int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, PSTR lpCmdLine, int nCmdShow) {
+                    return 0;
+                }
+                #else
+                // FUNCTION: {original_id} 0x10000020
+                int main(int argc, char *argv[]) {
+                    return 0;
+                }
+                #endif
+            """
+            )
+        case TargetType.SHARED_LIBRARY:
+            entry_function = textwrap.dedent(
+                """\
+                #ifdef _WIN32
+                #include <windows.h>
+
+                // FUNCTION: {original_id} 0x10000020
+                BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved ) {
+                    return TRUE;
+                }
+                #endif
+            """
+            )
+    return textwrap.dedent(
+        f"""\
+        #include "{hpp_path.name}"
+
+        // FUNCTION: {target_id} 0x10000000
+        SomeClass::~SomeClass() {{
+        }}
+
+        // GLOBAL: {target_id} 0x10102000
+        // STRING: {target_id} 0x10101f00
+        const char* g_globalString = "A global string";
+
+        {entry_function}
```

Comment:
I think this will work as:
```python
return textwrap.dedent("big string starting with #include") + entry_function
```
This string and `entry_function` should get dedented separately and then added together. In the current code, the top of the .cpp file keeps its indents. The lines from `entry_function` already have the padding removed, so we can't dedent further.

Reply:
I did something like that, with an extra newline for style :)